webapp/channels/src/actions/post_actions.ts

```diff
@@ -141,7 +147,28 @@ function storeCommentDraft(rootPostId: string, draft: null) {
 }
 
 export function addReaction(postId: string, emojiName: string) {
-    return (dispatch: DispatchFunc) => {
+    const getUniqueReactionsToPost = makeGetUniqueReactionsToPost();
+    return (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const state = getState() as GlobalState;
+        const config = getConfig(state);
+        const reactions = Object.values(getUniqueReactionsToPost(state, postId) ?? {});
+        let reactionCount = reactions.length;
+        if (reactions.some((reaction) => reaction.emoji_name === emojiName)) {
+            reactionCount--;
+        }
+
+        if (reactionCount >= Number(config.UniqueEmojiReactionLimitPerPost)) {
+            dispatch(openModal({
+                modalId: ModalIdentifiers.REACTION_LIMIT_REACHED,
+                dialogType: ReactionLimitReachedModal,
+                dialogProps: {
+                    isAdmin: isCurrentUserSystemAdmin(state),
+                    onExited: () => closeModal(ModalIdentifiers.REACTION_LIMIT_REACHED),
+                },
+            }));
+            return {data: false};
+        }
+
         dispatch(PostActions.addReaction(postId, emojiName));
         dispatch(addRecentEmoji(emojiName));
```

Comment:
We don't, but it's highly unlikely that they would see that. At that point the user would see a 400 error in the console if anything at all. We could pop the modal if we think that's a good idea (cc @abhijit-singh).

As for having the server drive it, I think it makes sense to avoid hitting the server with a request we know is going to fail if we can avoid it. Since we have the information on reactions on the client, it makes sense to use it there. The server can act as a failsafe.

Reply:
> As for having the server drive it, I think it makes sense to avoid hitting the server with a request we know is going to fail if we can avoid it. Since we have the information on reactions on the client, it makes sense to use it there. 

If we used that data to hide or disable the button in some way, I think I would agree, but I'm wondering if there's an opportunity to simplify here instead of effectively having to duplicate the logic twice.

Reply:
I think we're looking at two different paths altogether. We are adding some extra UX here for the main path (being a user clicks on a reaction and it's already at the limit) to support the main app case, and to avoid hitting the server at all. I suppose it could simplify things, but I think it's worth it to have the extra UX and the more optimized case of not forcing the server to make the decision. Plus this way our API works independently and we still have protection on it (fixing the original issue) to make sure it definitely cannot be forced.

Reply:
I admit I don't immediately see the need to "avoid hitting the server at all" (for what is already an extreme edge case), nor why this would change the needed API semantics. But I'm only 1/5 on this particular change on reducing complexity.

Reply:
I'll leave as-is for the time being.